src/cargo/ops/registry/login.rs

```diff
@@ -47,114 +33,21 @@ pub fn registry_login(
-            secret_key = Secret::default().map(|mut key| {
-                FormatAsPaserk::fmt(&kp.secret, &mut key).unwrap();
-                key
-            });
-        } else if secret_key_required {
-            assert!(!generate_keypair);
-            drop_println!(config, "please paste the API secret key below");
-            secret_key = Secret::default()
-                .map(|mut line| {
-                    let input = io::stdin();
-                    input
-                        .lock()
-                        .read_line(&mut line)
-                        .with_context(|| "failed to read stdin")
-                        .map(|_| line.trim().to_string())
-                })
-                .transpose()?;
-        } else {
-            secret_key = old_secret_key
-                .cloned()
-                .ok_or_else(|| anyhow!("need a secret_key to set a key_subject"))?;
-        }
-        if let Some(p) = paserk_public_from_paserk_secret(secret_key.as_deref()) {
-            drop_println!(config, "{}", &p);
-        } else {
-            bail!("not a validly formatted PASERK secret key");
-        }
-        new_token = RegistryCredentialConfig::AsymmetricKey((
-            secret_key,
-            match key_subject {
-                Some(key_subject) => Some(key_subject.to_string()),
-                None => old_key_subject,
-            },
-        ));
-    } else {
-        new_token = RegistryCredentialConfig::Token(match token {
-            Some(token) => token.owned(),
-            None => {
-                if let Some(login_url) = login_url {
-                    drop_println!(
-                        config,
-                        "please paste the token found on {} below",
-                        login_url
-                    )
-                } else {
-                    drop_println!(
-                        config,
-                        "please paste the token for {} below",
-                        source_ids.original.display_registry_name()
-                    )
-                }
 
-                let mut line = String::new();
-                let input = io::stdin();
-                input
-                    .lock()
-                    .read_line(&mut line)
-                    .with_context(|| "failed to read stdin")?;
-                // Automatically remove `cargo login` from an inputted token to
-                // allow direct pastes from `registry.host()`/me.
-                Secret::from(line.replace("cargo login", "").trim().to_string())
-            }
-        });
-
-        if let Some(tok) = new_token.as_token() {
-            crates_io::check_token(tok.as_ref().expose())?;
+    let mut token_from_stdin = None;
+    if !std::io::stdin().is_terminal() {
+        let token = std::io::read_to_string(std::io::stdin()).unwrap_or_default();
+        if !token.is_empty() {
+            token_from_stdin = Some(token);
         }
     }
```

Comment:
Just a friendly heads up - this is causing our CI jobs to hang if we provide the token via the command-line instead of `stdin`.

Reply:
Sorry for the inconvenience!

I didn't quite follow the PR at that time. Is that a regression? If yes, please file an issue with reproduction, and we can then arrange a fix!

Reply:
Yes, this is a regression.
Sounds good, thanks :)

Reply:
#12438